### ISSUE TITLE
Speck, Android and Linux kernel interop

### ISSUE THREAD

**Issue by noloader** [weidai11/cryptopp#585]
The Linux kernel added Speck-64 and Speck-128 for opportunistic encryption on low resource devices that would otherwise have to disable encryption on the file system. Also see [[PATCH v2 0/5] crypto: Speck support](https://www.mail-archive.com/linux-crypto@vger.kernel.org/msg30925.html) on the kernel-crypto mailing list.

The problem is we are not going to interop well with the kernel. The problem is due to confusion in the Simon and Speck algorithmic description and the test vectors. The algorithmic description and test vectors did not quite align, and we went down one path (follow the test vectors) and the kernel went down another path (follow the algorithmic description).

It is possible to to interop with the kernel using our implementation but... (1) it is not readily apparent how to do it, and (2) there's a loss of efficiency when doing it. The interop can happen now by providing an adapter class. If my estimates are correct, then we could loose 2 to 4 cpb on a conversion via the adapter. That means SSE4 could drop from 2.1 cpb to about 4 or 6 cpb.

Also not readily apparent, the same interop issue affect Simon, too. At this point in time we are not aware of anyone implementing Simon and diverging due to the algorithmic description versus test vector results.

Simon and Speck has been part of the library since December 2017. We are probably running between the cracks, but we can probably make this change and release 6.1 before anyone notices. The existing classes will be modified to align with the kernel at speed. An adapter class can be used for the old [incorrect] behavior if needed. Thos who want the old behavior will pay the performance penalty.

This will track our handling of the issue and document how we close the gap.

-----

Also see [Speck, Android and Linux kernel interop](https://groups.google.com/d/msg/cryptopp-users/urd3pcd2ees/ojHnONPzAgAJ) on the Crypto++ mailing list.

**Comment by ebiggers** [comment]
> It is possible to to interop with the kernel using our implementation but... (1) it is not readily apparent how to do it, and (2) there's a loss of efficiency when doing it. If my estimates are correct, then we could loose 2 to 4 cpb on a conversion. That means SSE4 could drop from 2.1 cpb to about 4 or 6 cpb.

Why would it be slower?  The data only needs to be packed/unpacked once, not once per cipher round.  With SSE2 you need  _mm_unpacklo_epi64() (PUNPCKLQDQ) and _mm_unpackhi_epi64 (PUNPCKHQDQ) anyway to de-interleave the (x, y), and similarly for interleaving them again at the end.  So it should be easy to relabel the registers such that the word order is (y, x).  Also since the byte order is now little endian the byte swap via _mm_shuffle_epi8(), i.e. PSHUFB, isn't needed anymore which would make the new convention faster actually.

Also just to clarify, the above linked LKML thread is for v1 of the patchset which used the (x, y) word order.  But in v2 (https://www.mail-archive.com/linux-crypto@vger.kernel.org/msg30925.html) it has been changed to (y, x) since that's what the Speck designers said is the intended order.

**Comment by noloader** [comment]
@ebiggers,

> Why would it be slower? The data only needs to be packed/unpacked once, not once per cipher round

My bad... I was referring to an adapter class to frame the data. That would accommodate both behaviors with minimal changes. Let me get that added above.

The other options are (1) add a second implementation so both behaviors are stand alone and perform at speed, (2) provide only one implementation and align with the kernel's implementation by modifying the existing implementation. A variation on (2) is, provide the adapter class on top of the updated Speck implementation so the old behavior is still available through the adapter but slower.

I'm learning towards (2). We'll modify the Crypto++ implementation to align with the kernel's implementation. The modified Speck will interop and run at speed. Those who need the incorrect behavior can get it through an adapter class that runs a little slower.

> Also just to clarify, the above linked LKML thread is for v1 of the patchset which used the (x, y) word order. But in v2 ... it has been changed to (y, x) since that's what the Speck designers said is the intended order.

OK, thanks. The reference has been changed.

**Comment by noloader** [comment]
Simon and Speck were removed at [Commit 15b14cc61890](https://github.com/weidai11/cryptopp/commit/15b14cc61890). This is a stop-gap measure to ensure we don't do any harm.

The original files for the Simon and Speck implementation are available in `simon-speck.zip`. They are in GitHub for folks who need them.

We will loop back around to this issue when we have the time to untangle things. We really need (1) a good set of test vectors; and (2) a working reference implementation. We don't have either at the moment.

We have one test vector from the Linux kernel, but our implementation failed to arrive at the known answer. We did some knob turning but failed to arrive at the answer.

Here is the kernel test vector provided by email. More should be available in the patch set for `testmgr.c` on the kernel mailing list. Hopefully the Simon and Speck authors will provide some working test vectors so we don't have to pull them out of mailing lists posts.

```
        {
                .key    = "\x00\x01\x02\x03\x04\x05\x06\x07"
                          "\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f"
                          "\x10\x11\x12\x13\x14\x15\x16\x17"
                          "\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f",
                .klen   = 32,
                .input  = "\x70\x6f\x6f\x6e\x65\x72\x2e\x20"
                          "\x49\x6e\x20\x74\x68\x6f\x73\x65",
                .ilen   = 16,
                .result = "\x43\x8f\x18\x9c\x8d\xb4\xee\x4e"
                          "\x3e\xf5\xc0\x05\x04\x01\x09\x41",
                .rlen   = 16,
        },
```
